corehq/messaging/fcm_util.py

```diff
@@ -0,0 +1,74 @@
+import logging
+import os
+
+from django.conf import settings
+from firebase_admin import messaging, credentials, initialize_app
+
+logger = logging.getLogger(__name__)
+
+MAX_DEVICES_ALLOWED_MULTICAST = 500
+HQ_FCM_UTIL = None
+
+
+def is_fcm_available():
+    if not settings.FCM_CREDS_PATH:
+        return False
+    return os.path.isfile(settings.FCM_CREDS_PATH)
+
+
+class FCMUtilException(Exception):
+    pass
+
+
+class DevicesLimitExceeded(FCMUtilException):
+    def __init__(self):
+        super().__init__(f"Max devices allowed is {MAX_DEVICES_ALLOWED_MULTICAST}! Please execute in batches.")
+
+
+class FCMUtil:
+    def __init__(self, app_name, creds_path=settings.FCM_CREDS_PATH):
+        creds = credentials.Certificate(creds_path)
+        self.app = initialize_app(credential=creds, name=app_name)
+
+    @staticmethod
+    def _build_notification(title, body):
+        return messaging.Notification(
+            title=title,
+            body=body,
+        )
+
+    def send_to_single_device(self, registration_token, title, body, data=None):
+        """
+        Sends message to a single device.
+        https://firebase.google.com/docs/cloud-messaging/send-message#send-messages-to-specific-devices
+        """
+        message = messaging.Message(
+            notification=self._build_notification(title, body),
+            token=registration_token,
+            data=data
+        )
+        response = messaging.send(message, app=self.app)
+        # print(response)
+        return response
+
+    def send_to_multiple_devices(self, registration_tokens, title, body, data=None):
```

Comment:
Can we keep only this implementation and use it to send to single device as well?

Reply:
FYI : Both methods use different types of Firebase Message class and behave differently under the hood with different responses as well.
Have kept both the methods available and delegated the decision to use either of them to the Consumer of this class.